In [1]:
import pandas as pd
import numpy as np
from sqlalchemy import create_engine


In [2]:
engine = create_engine(
    
     "postgresql://postgres:postgres123@localhost:5432/malaria_db"
)

In [3]:
query = "SELECT * FROM malaria_cleaned"

df = pd.read_sql(query, engine)

In [4]:
feature_df = df.copy()

In [5]:
feature_df = df.copy()
print(df.shape)

(730, 9)


# Total_doses

 purpose: Measures participant treatment exposure and adherence.
 
 Logic: Count all dose records per participant.

In [6]:
df['total_doses'] = (
    df.groupby('participant_id')
      ['dose_num']
      .transform('count')
)

In [7]:
#verification
df[['participant_id','total_doses']].head()

,participant_id,total_doses
0,1,4
1,1,4
2,1,4
3,1,4
4,2,4


In [8]:
df['total_doses'].describe()

count    730.000000
mean       3.906849
std        0.794067
min        1.000000
25%        4.000000
50%        4.000000
75%        4.000000
max        8.000000
Name: total_doses, dtype: float64

The feature was successfully generated using participant-level aggregation.

Each participant was assigned the total number of doses received during the treatment period, with the value repeated across all corresponding dose records.

This feature will support adherence analysis, protocol completion assessment, and treatment exposure KPIs.


In [9]:
participant_summary = (
    df.groupby('participant_id')
      .agg(
          total_doses=('dose_num','count')
      )
)

participant_summary['total_doses'].value_counts().sort_index()


total_doses
1      5
2      3
3     39
4    145
6      1
8      2
Name: count, dtype: int64

In [10]:
#verification
df[['participant_id','dose_num','total_doses']].head(10)

,participant_id,dose_num,total_doses
0,1,1,4
1,1,2,4
2,1,3,4
3,1,4,4
4,2,1,4
5,2,2,4
6,2,3,4
7,2,4,4
8,3,1,4
9,3,2,4


# Observation
Verification of the `total_doses` feature confirmed that participant-level treatment exposure was calculated correctly.
Among the 195 participants, 145 received the standard four-dose regimen, while 47 received fewer than four doses and 3 received more than four doses. 
The feature successfully captures treatment exposure and will serve as a foundational variable for adherence analysis, protocol completion measurement, and treatment exposure KPIs.


# Completed_protocol
Purpose: Identifies participants who completed the expected regimen


In [11]:
df['completed_protocol']= (
df['total_doses']>=4
).astype(int)

In [12]:
df['completed_protocol'].value_counts()

completed_protocol
1    602
0    128
Name: count, dtype: int64

In [13]:
# Participant_level verification
(
    df.groupby('participant_id')['completed_protocol']
      .first()
      .value_counts()
)

completed_protocol
1    148
0     47
Name: count, dtype: int64

# Observation
Verification confirmed that the completed_protocol feature was generated correctly. Participants with four or more recorded doses were flagged as protocol completers.
At the participant level, 148 individuals met the completion criteria, while 47 participants received fewer than four doses.
This feature will be used to calculate treatment completion rates and assess adherence to the expected treatment regimen.


# Incomplete_protocol
Identifies participants receiving fewer doses than expected

In [14]:
df['incomplete_protocol']= (
df['total_doses']<4
).astype(int)

In [15]:
(
    df.groupby('participant_id')['incomplete_protocol']
      .first()
      .value_counts()
)

incomplete_protocol
0    148
1     47
Name: count, dtype: int64

# Observation
47 participants received fewer than 4 doses while 148 individuals met the compeletion criteria
This feature will be used to calculate treatment incompletion rates.

# Extended_treatment
Detects re_treatment or protocol deviations

In [16]:
df['extended_treatment']= (
df['total_doses']>4
).astype(int)

In [17]:
df['extended_treatment'].value_counts()

extended_treatment
0    708
1     22
Name: count, dtype: int64

In [18]:
#participant_level verificatin
df.loc[
    df['extended_treatment'] == 1,
    ['participant_id', 'total_doses']
].drop_duplicates()

,participant_id,total_doses
416,108,8
541,142,8
626,164,6


In [19]:
df[
    (df['extended_treatment'] == 1)
    & (df['total_doses'] <= 4)
]

,participant_id,dose_num,dose_ordered,dose_start,dose_start_time,dose_end,dose_end_time,dose_complete,treatment_duration_days,total_doses,completed_protocol,incomplete_protocol,extended_treatment


# Observation
Verification confirmed that the `extended_treatment` feature was generated correctly. Participants with more than four recorded doses were flagged as receiving extended treatment exposure. 
Three participants met this criterion, corresponding to previously identified cases of repeated treatment cycles or protocol deviations. 
This feature will support the calculation of extended treatment rates and the identification of atypical treatment pathways.


# total_exposure
Measures cumulative medication exposure.

In [20]:
df['total_exposure']= (
df.groupby('participant_id')
      ['dose_ordered']
.transform('sum')
)

In [21]:
df[['participant_id',
'dose_ordered',
'total_exposure']]

,participant_id,dose_ordered,total_exposure
0,1,250.0,1000.0
1,1,250.0,1000.0
2,1,250.0,1000.0
3,1,250.0,1000.0
4,2,NaN,660.0
...,...,...,...
725,196,168.0,504.0
726,196,168.0,504.0
727,197,120.0,360.0
728,197,120.0,360.0


In [22]:
df['dose_ordered'].isna().sum()

np.int64(1)

In [23]:
#summary statistics
df['total_exposure'].describe()

count     730.000000
mean      633.069725
std       270.669792
min        94.000000
25%       440.000000
50%       652.000000
75%       832.000000
max      1444.000000
Name: total_exposure, dtype: float64

Verification confirmed that the `total_exposure` feature was generated correctly by summing all ordered doses for each participant across the treatment period.
The resulting cumulative exposure value was repeated across all records belonging to the same participant.
This feature provides a participant-level measure of overall medication exposure and will support treatment intensity analysis, exposure monitoring, and healthcare KPI development.


# Highest expsure participants
Purpose: This helps identify participants with the greatest cumulative medication exposure.

In [24]:
(
    df.groupby('participant_id')
      ['total_exposure']
      .first()
      .sort_values(ascending=False)
      .head(10)
)

participant_id
142    1444.000000
147    1200.000000
148    1200.000000
27     1163.599976
84     1132.000000
74     1120.000000
87     1088.000000
192    1080.000000
129    1023.359985
6      1000.000000
Name: total_exposure, dtype: float64

# Insight
1.  Verification of the `total_exposure` feature identified substantial variation in cumulative medication exposure across participants.
2. Participant 142 exhibited the highest exposure value (1444 units), consistent with previously identified repeated treatment cycles. 
3. Several participants following standard treatment schedules also demonstrated high cumulative exposure, indicating variability in dose quantities administered.
4. The feature successfully captures participant-level treatment exposure and will support exposure monitoring, treatment intensity assessment, and cumulative exposure KPIs.


# statistic summary

In [25]:
(
    df.groupby('participant_id')
      ['total_exposure']
      .first()
      .describe()
)


count     195.000000
mean      612.260820
std       266.262472
min        94.000000
25%       394.000000
50%       624.000000
75%       820.000000
max      1444.000000
Name: total_exposure, dtype: float64

1. Summary statistics for the `total_exposure` feature confirmed substantial variation in cumulative medication exposure across participants.
2. The average exposure was 612.3 units, with a median of 624 units, indicating a relatively balanced distribution. Exposure values ranged from 94 to 1444 units, reflecting differences in dose quantities administered and treatment pathways.
3. The feature successfully captures participant-level medication exposure and will support exposure monitoring, treatment intensity analysis, and cumulative exposure KPI development.


# Avg_dose_ordered
Purpose : Measures average treatment intensity

In [26]:
df['avg_dose_ordered']= (
df.groupby('participant_id')
      ['dose_ordered']
.transform('mean')
)

In [27]:
#verification
df['avg_dose_ordered'].describe()

count    730.000000
mean     163.850493
std       65.283707
min       23.500000
25%      120.000000
50%      173.250000
75%      210.000000
max      360.000000
Name: avg_dose_ordered, dtype: float64

# Insight
Verification of the `avg_dose_ordered` feature confirmed meaningful variation in medication dosing across the study population. 
The average dose ordered was 163.9 units, with most values falling between 120 and 210 units.
The observed variability suggests individualized dosing practices rather than a uniform fixed-dose regimen. This feature will support treatment intensity analysis, exposure monitoring, and dose-related healthcare KPI development.


In [28]:
# Participant level verificatin
(
    df.groupby('participant_id')
      ['avg_dose_ordered']
      .first()
      .describe()
)

count    195.000000
mean     166.007316
std       65.830085
min       23.500000
25%      125.573334
50%      174.000000
75%      212.166667
max      360.000000
Name: avg_dose_ordered, dtype: float64

# Observation
1. Verification of the avg_dose_ordered feature confirmed substantial variation in participant-level medication dosing.
2. The mean average dose was 166.0 units, with a median of 174.0 units, suggesting a relatively balanced distribution.
3. Average doses ranged from 23.5 to 360 units, while 50% of participants received average doses between 125.6 and 212.2 units.

 These findings indicate that medication dosing was individualized across participants and support the use of this feature for treatment intensity assessment, exposure monitoring, and dose-related KPI development.

# Avg_duration
Purpose : Measures participant-level treatment timing.

In [29]:
df['avg_duration']= (
df.groupby('participant_id')
      ['treatment_duration_days']
.transform('mean')
)

In [30]:
#verification
df['avg_duration'].describe()

count    588.000000
mean       0.045918
std        0.414824
min        0.000000
25%        0.000000
50%        0.000000
75%        0.000000
max        5.000000
Name: avg_duration, dtype: float64

verification of the avg_duration feature confirmed that treatment duration was minimal across the cohort.

The mean participant-level duration was approximately 0.05 days, while the median duration was 0 days, indicating that most treatments were completed on the same day.

Duration variability was limited and driven by a small number of extended-duration records.

This feature will support treatment timing analysis and duration-related operational KPIs.

In [31]:
# Participant level verification
(
    df.groupby('participant_id')
      ['avg_duration']
      .first()
      .describe()
)

count    158.000000
mean       0.042722
std        0.401225
min        0.000000
25%        0.000000
50%        0.000000
75%        0.000000
max        5.000000
Name: avg_duration, dtype: float64

In [32]:
(
    df.groupby('participant_id')
      ['avg_duration']
      .first()
      .isna()
      .sum()
)


np.int64(37)

# Observation
Verification of the avg_duration feature showed that average treatment duration could be calculated for 158 participants. 

The mean duration was 0.04 days, while the median duration was 0 days, indicating that treatment administration was typically completed on the same day.

Duration variability was minimal and largely driven by a small number of participants with extended treatment periods.

Due to the high concentration of zero values and missing duration information, this feature is expected to play a supporting role in KPI development rather than serve as a primary performance metric.

# max_duration
Purpose : Detects participant_level anomalies

In [33]:
df['max_duration']= (
df.groupby('participant_id')
      ['treatment_duration_days']
.transform('max')
)

In [34]:
df['max_duration'].describe()

count    588.000000
mean       0.183673
std        1.659295
min        0.000000
25%        0.000000
50%        0.000000
75%        0.000000
max       20.000000
Name: max_duration, dtype: float64

Verification of the max_duration feature demonstrated that most participants experienced no measurable treatment duration beyond the day of administration. 

However, a small number of participants exhibited extended treatment durations, including one extreme case of 20 days.

The feature successfully captures participant-level duration anomalies and will support identification of prolonged treatment events, documentation reviews, and duration-related quality monitoring.

In [35]:
# Participant_level verification
(
    df.groupby('participant_id')
      ['max_duration']
      .first()
      .describe()
)

count    158.000000
mean       0.170886
std        1.604901
min        0.000000
25%        0.000000
50%        0.000000
75%        0.000000
max       20.000000
Name: max_duration, dtype: float64

In [36]:
(
    df.groupby('participant_id')
      ['max_duration']
      .first()
      .sort_values(ascending=False)
      .head(10)
)

participant_id
68     20.0
63      2.0
88      1.0
37      1.0
5       1.0
18      1.0
133     1.0
8       0.0
10      0.0
9       0.0
Name: max_duration, dtype: float64

# Observation
Verification of the max_duration feature confirmed that treatment duration was minimal for most participants.

Among the 158 participants with available duration information, at least 75% had a maximum duration of zero days.

Only seven participants experienced treatment durations greater than zero days, with participant 68 exhibiting an extreme duration of 20 days. 

This feature effectively identifies prolonged treatment events and duration anomalies that may require further investigation for operational or data-quality purposes.

# missing_end_date
Purpose : Measures documentation completeness.

In [37]:
df['missing_end_date']= (
df['dose_end']
.isna()
.astype(int)
)

In [38]:
df['missing_end_date'].value_counts()

missing_end_date
0    541
1    189
Name: count, dtype: int64

# Observation
Verification of the missing_end_date feature confirmed that 189 treatment records (25.9%) lacked a documented end date, while 541 records (74.1%) contained complete end-date information.

Missing end dates were directly associated with missing treatment duration values, highlighting an important data-quality issue affecting duration-based analyses.

This feature will support documentation completeness monitoring and data-quality KPI development.

# missing_duration
Purpose : Measures availability of duration information.

In [39]:
df['missing_duration']= (
df['treatment_duration_days']
.isna()
.astype(int)
)

In [40]:
df['missing_duration'].value_counts()

missing_duration
0    541
1    189
Name: count, dtype: int64

# Observation
Verification of the missing_duration feature confirmed that 189 treatment records (25.9%) lacked duration information, while 541 records (74.1%) contained complete duration data. 

The missing duration values correspond exactly to records with missing treatment end dates, indicating that duration incompleteness was driven by documentation gaps rather than treatment status.

This feature will support data-quality monitoring and duration-data completeness KPIs.

# Data exportatinon

In [41]:
df.to_sql(
    "malaria_feature_engineered",
    engine,
    if_exists="replace",
    index=False
)


pd.read_sql(
    "SELECT * FROM malaria_feature_engineered LIMIT 5",
    engine
)

,participant_id,dose_num,dose_ordered,dose_start,dose_start_time,dose_end,dose_end_time,dose_complete,treatment_duration_days,total_doses,completed_protocol,incomplete_protocol,extended_treatment,total_exposure,avg_dose_ordered,avg_duration,max_duration,missing_end_date,missing_duration
0,1,1,250.0,2019-04-05,TRUE,2019-04-05,TRUE,1,0.0,4,1,0,0,1000.0,250.0,0.0,0.0,0,0
1,1,2,250.0,2019-04-06,TRUE,2019-04-06,TRUE,1,0.0,4,1,0,0,1000.0,250.0,0.0,0.0,0,0
2,1,3,250.0,2019-04-06,TRUE,2019-04-06,TRUE,1,0.0,4,1,0,0,1000.0,250.0,0.0,0.0,0,0
3,1,4,250.0,2019-04-07,TRUE,2019-04-07,TRUE,1,0.0,4,1,0,0,1000.0,250.0,0.0,0.0,0,0
4,2,1,NaN,2019-04-11,TRUE,2019-04-11,FALSE,1,0.0,4,1,0,0,660.0,220.0,0.0,0.0,0,0


## Participant_dashboard

In [42]:
participant_dashboard = (
    df.groupby('participant_id')
      .agg(
          total_doses=('dose_num', 'count'),
          total_exposure=('dose_ordered', 'sum'),
          avg_dose_ordered=('dose_ordered', 'mean'),
          avg_duration=('treatment_duration_days', 'mean'),
          max_duration=('treatment_duration_days', 'max'),
          completed_protocol=('completed_protocol', 'max'),
          incomplete_protocol=('incomplete_protocol', 'max'),
          missing_end_date=('missing_end_date', 'max'),
          missing_duration=('missing_duration', 'max'),
          
          # Add the missing column here:
          extended_treatment=('extended_treatment', 'max') 
      )
      .reset_index()
)

In [43]:
participant_dashboard.head()

participant_dashboard.shape

participant_dashboard.describe()

,participant_id,total_doses,total_exposure,avg_dose_ordered,avg_duration,max_duration,completed_protocol,incomplete_protocol,missing_end_date,missing_duration,extended_treatment
count,195.000000,195.00000,195.000000,195.000000,158.000000,158.000000,195.000000,195.000000,195.000000,195.000000,195.000000
mean,99.184615,3.74359,612.260820,166.007316,0.042722,0.170886,0.758974,0.241026,0.338462,0.338462,0.015385
std,57.172706,0.78379,266.262472,65.830085,0.401225,1.604901,0.428807,0.428807,0.474404,0.474404,0.123394
min,1.000000,1.00000,94.000000,23.500000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,50.500000,4.00000,394.000000,125.573334,0.000000,0.000000,1.000000,0.000000,0.000000,0.000000,0.000000
50%,99.000000,4.00000,624.000000,174.000000,0.000000,0.000000,1.000000,0.000000,0.000000,0.000000,0.000000
75%,148.500000,4.00000,820.000000,212.166667,0.000000,0.000000,1.000000,0.000000,1.000000,1.000000,0.000000
max,197.000000,8.00000,1444.000000,360.000000,5.000000,20.000000,1.000000,1.000000,1.000000,1.000000,1.000000


## Export


In [44]:
participant_dashboard.to_sql(

    'participant_dashboard',

    engine,

    if_exists='replace',

    index=False

)

195